storage/fcds/fcds.go

```diff
@@ -0,0 +1,358 @@
+	wg           sync.WaitGroup // blocks Close until all other method calls are done
+	maxChunkSize int            // maximal chunk data size
+	quit         chan struct{}  // quit disables all operations after Close is called
+	quitOnce     sync.Once      // protects quit channel from multiple Close calls
+}
+
+// Option is an optional argument passed to New.
+type Option func(*Store)
+
+// WithCache is an optional argument to New constructor that enables
+// in memory cache of free chunk data positions in files
+func WithCache(yes bool) Option {
+	return func(s *Store) {
+		if yes {
+			s.freeCache = newOffsetCache(shardCount)
+		} else {
+			s.freeCache = nil
+		}
+	}
+}
+
+// New constructs a new Store with files at path, with specified max chunk size.
+func New(path string, maxChunkSize int, metaStore MetaStore, opts ...Option) (s *Store, err error) {
+	s = &Store{
+		shards:       make([]shard, shardCount),
+		meta:         metaStore,
+		free:         make([]bool, shardCount),
+		maxChunkSize: maxChunkSize,
+		quit:         make(chan struct{}),
+	}
+	for _, o := range opts {
+		o(s)
+	}
+	if err := os.MkdirAll(path, 0777); err != nil {
+		return nil, err
+	}
+	for i := byte(0); i < shardCount; i++ {
+		s.shards[i].f, err = os.OpenFile(filepath.Join(path, fmt.Sprintf("chunks-%v.db", i)), os.O_CREATE|os.O_RDWR, 0666)
+		if err != nil {
+			return nil, err
+		}
+		s.shards[i].mu = new(sync.Mutex)
+	}
+	return s, nil
+}
+
+// Get returns a chunk with data.
+func (s *Store) Get(addr chunk.Address) (ch chunk.Chunk, err error) {
+	if err := s.protect(); err != nil {
+		return nil, err
+	}
+	defer s.unprotect()
+
+	sh := s.shards[getShard(addr)]
+	sh.mu.Lock()
+	defer sh.mu.Unlock()
+
+	m, err := s.getMeta(addr)
+	if err != nil {
+		return nil, err
+	}
+	data := make([]byte, m.Size)
+	n, err := sh.f.ReadAt(data, m.Offset)
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+	if n != int(m.Size) {
+		return nil, fmt.Errorf("incomplete chunk data, read %v of %v", n, m.Size)
+	}
+	return chunk.NewChunk(addr, data), nil
+}
+
+// Has returns true if chunk is stored.
+func (s *Store) Has(addr chunk.Address) (yes bool, err error) {
+	if err := s.protect(); err != nil {
+		return false, err
+	}
+	defer s.unprotect()
+
+	mu := s.shards[getShard(addr)].mu
+	mu.Lock()
+	defer mu.Unlock()
+
+	_, err = s.getMeta(addr)
+	if err != nil {
+		if err == chunk.ErrChunkNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// Put stores chunk data.
+func (s *Store) Put(ch chunk.Chunk) (err error) {
+	if err := s.protect(); err != nil {
+		return err
+	}
+	defer s.unprotect()
+
+	addr := ch.Address()
+	data := ch.Data()
+
+	section := make([]byte, s.maxChunkSize)
+	copy(section, data)
+
+	shard := getShard(addr)
+	sh := s.shards[shard]
+
+	sh.mu.Lock()
+	defer sh.mu.Unlock()
+
+	offset, reclaimed, err := s.getOffset(shard)
+	if err != nil {
+		return err
+	}
+
+	if offset < 0 {
+		offset, err = sh.f.Seek(0, io.SeekEnd)
```

Comment:
getOffset comment does say what the return value is from that method and it is not its responsibility to declare what is done in other methods. I thought that it is clear that it would be appended. I will add more explanation for this behaviour.